solidpy/Burn.py

```diff
@@ -80,105 +190,99 @@ def vector_field(self, time, state_variables):
-            k * R * self.evaluate_exit_temperature()
+        (
+            self.initial_tail_off_time,
+            self.initial_tail_off_chamber_pressure,
+            self.initial_tail_off_free_volume,
+        ) = (self.solution.t[-1], self.solution.y[0][-1], self.solution.y[1][-1])
+
+        self.evaluate_tail_off_chamber_pressure = (
+            lambda time: self.initial_tail_off_chamber_pressure
+            * math.exp(
+                -R
+                * T_0
+                * A_t
+                / (self.initial_tail_off_free_volume * self.propellant.calc_cstar())
+                * (time - self.initial_tail_off_time)
+            )
         )
-        return self.exit_velocity
 
-    def evaluate_Cf(self):
-        _, _, _, k, _ = self.parameters
-        self.Cf = math.sqrt(
-            (2 * k / (k - 1))
-            * math.pow(2 / (k + 1), (k + 1) / (k - 1))
-            * (1 - math.pow(self.exit_pressure / self.chamber_pressure, (k - 1) / k))
-        )
-        return self.Cf
+        self.tail_off_time = []
+        self.tail_off_chamber_pressure = []
 
-    def evaluate_thrust(self, chamber_pressure):
-        self.thrust = (
-            self.evaluate_nozzle_mass_flow(chamber_pressure)
-            * self.evaluate_exit_velocity()
-            + (self.evaluate_exit_pressure(chamber_pressure) - self.initial_pressure)
-            * self.motor.nozzle_exit_area
-        )
-        return self.thrust
+        for time in np.linspace(*self.time_span, self.number_steps + 1):
+            tail_off_chamber_pressure = self.evaluate_tail_off_chamber_pressure(
+                time + self.initial_tail_off_time
+            )
+            if tail_off_chamber_pressure / self.environment_pressure < 1.0001:
+                break
+            else:
+                self.tail_off_time.append(time + self.initial_tail_off_time)
+                self.tail_off_chamber_pressure.append(tail_off_chamber_pressure)
 
-    def evaluate_thrust2(self):
-        self.evaluate_Cf()
-        self.thrust = self.Cf * self.chamber_pressure * self.motor.nozzle_throat_area
-        return self.thrust
+        self.tail_off_solution = [self.tail_off_time, self.tail_off_chamber_pressure]
+
+        return self.tail_off_solution
 
-    def evaluate_specific_impulse(self, chamber_pressure_list, thrust_list):
-        self.specific_impulse_list = []
-        for chamber_pressure, thrust in zip(chamber_pressure_list, thrust_list):
-            self.specific_impulse_list.append(
-                thrust / (self.evaluate_nozzle_mass_flow(chamber_pressure))
-            )
-        specific_impulse = np.average(self.specific_impulse_list) / self.gravity
-        return specific_impulse
 
-    def empirical_evaluate_chamber_pressure(self):
+class BurnEmpirical(Burn):
+    def __init__(
+        self,
+        grain,
+        motor,
+        propellant,
+        environment=Environment(),
+        empirical_data=None,
+    ):
```

Comment:
Have you considered this class should perhaps be moved over into a different file? IMHO, it would be nice to have a complete and robust set of "Empirical Analysis" features from static fire data (and lots of other stuff which could be contemplated in post-analysis). 

Maybe it's enough to justify a class of its own. Plus, it would have the added benefit of uncluttering a bit the main simulation file (considering no data available a priori).

Reply:
I agree that it would be more interesting to have a separate file for "Empirical Analysis" and it would be rather convenient to unclutter the main simulation file.

However, there are many ways we can do it regarding class structure, because currently burn_empirical is a subclass of Burn and has a strong dependency of its __init__. Thus, I am open for suggestions about its functioning and structure. Personally, I would keep it as a subclass of Burn and simply move it to a different file.

Reply:
Keeping the inheritance between both classes, but moving the child `BurnEmpirical` over to a different file seems like a neat solution to me!

i see you've already made the respective changes. Good work!